### ISSUE TITLE
Deprecate `insert_zero` (and thus `gap_policy`) on pipeline aggs

### ISSUE THREAD

**Issue by polyfractal** [elastic/elasticsearch#24893]
The `insert_zero` gap policy has bothered me for a long time, and a related discussion in #24747 made me think about it again.  Inserting zeros leads to all kinds of weird responses, like this:

```json
{
  "aggregations": {
    "histo": {
      "buckets": [
        {
          "key_as_string": "2015-01-03T00:00:00.000Z",
          "key": 1420243200000,
          "doc_count": 1,
          "max1": {
            "value": 1
          },
          "movavg": {
            "value": 1
          }
        },
        {
          "key_as_string": "2015-01-04T00:00:00.000Z",
          "key": 1420329600000,
          "doc_count": 1,
          "max1": {
            "value": null
          },
          "movavg": {
            "value": 1
          }
        },
        {
          "key_as_string": "2015-01-05T00:00:00.000Z",
          "key": 1420416000000,
          "doc_count": 1,
          "max1": {
            "value": null
          },
          "movavg": {
            "value": 0.6666666666666666
          }
        },
        {
          "key_as_string": "2015-01-06T00:00:00.000Z",
          "key": 1420502400000,
          "doc_count": 1,
          "max1": {
            "value": null
          },
          "movavg": {
            "value": 0.3333333333333333
          }
        }
      ]
    }
  }
}
```

Where the value is presented to the user as `null`, but the movavg magically keeps generating values.  This is at best confusing.  Philosophically, it's dangerous in that it's doctoring the data we return to the user.  For the same reasons I argued in #24747, if used as part of an "analytics pipeline", `insert_zeros` will generate unsound results because it's just making up data... and in an astoundingly unintelligent manner at that.

I vote we deprecate `insert_zero` and by extension all of `gap_policy` immediately, and remove it at a later date.  This also has the nice benefit of simplifying the internal pipeline code a bit, since the gap_policy stuff is kinda kludgy anyway.

I did a search through the issues and forums, and there's essentially no user discussion of the option being used (which is probably a good thing).

/cc @colings86 @jpountz

**Comment by mpaduada** [comment]
Whoa, whoa! Hold on. While I agree that `insert_zeros` is generally a bad idea, I think a parameter to specify some kind of `gap_policy` is still needed.

Missing/censored data is a perfectly normal occurrence when doing statistical analysis; accordingly statisticians employ different techniques to deal with missing data as dictated by the context. In financial time series analysis, it's quite common to be missing observations due to weekends and holidays; the missing data doesn't preclude calculating SMA, EMA, and other fun stats used in econometrics and technical analysis. The `skip` policy leaves you without a metric for a given bucket, which is essentially saying, "I don't know how to deal with this, so I'll pretend there's nothing to do."

Instead, I propose that options be added to `gap_policy`. Keep `skip`; it's perfectly reasonable in some cases. A `next_bucket` option could indicate that the next non-empty bucket be substituted; this is essentially the standard treatment for financial market data with missing values for weekends/holidays to calculate `serial_diff` and `moving_avg`. `next_bucket` would also be a nice option to have for `cumulative_sum`, which oddly doesn't accept a `gap_policy` parameter (even though the [documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-pipeline.html#gap-policy) indicates that all pipeline aggregations accept that parameter).  An `omit` bucket that calculates the metric using the available non-empty buckets (adjusting the denominator for `moving_avg`) would be a good replacement for the `insert_zeros` bucket. An `EM` option to use the [expectation-maximization algorithm](https://en.wikipedia.org/wiki/Expectation%E2%80%93maximization_algorithm) would likely please statisticians. And I'm sure statisticians and others with more complicated logging metrics use cases can weigh in with other suggested options.

As an aside, I suspect the reason that there's not a lot of discussion on `gap_policy` at this time is because some of the more complicated logging metrics use cases are only just starting to come on board to Elasticsearch. (I had no idea what Elasticsearch was three months ago.) I can personally attest that obtaining the utilization and efficiency metrics we need requires some aggregation acrobatics (though I might have an easier time if the documentation for [Scripted Metric Aggregation](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics-scripted-metric-aggregation.html) and [Painless](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-scripting-painless.html) were improved, or at least more numeric examples were included... but that's another rant).

TL; DR: Keep the `gap_policy` parameter and consider the inclusion of additional options.

**Comment by polyfractal** [comment]
Thanks for the feedback @mpaduada!  Really appreciate the input. :)

I'm curious how any kind of gap policy can lead to proper analysis though.  After thinking about it, even `skip` is looking wrong to me right now.  It seems to me that the only really useful policy would be `insert_NaN`, e.g. treat all gaps as true gaps/null/NaN.  Then gaps can then be cleaned up at the end of the analysis.

E.g considering financial data like you said, the weekends are usually missing.  If any kind of interpolating policy is used, a daily derivative will show a change from Friday to Saturday, which is just wrong.  Markets were closed, there was no change.  That seems to argue for some kind of `last_known_value` policy, but that's deceitful too.  It's not that there wasn't a change, it's that there was no collected data.  So even a derivative of `0` isn't true, the derivative is really `null/NaN`

Of course, NaNs/gaps on a chart are gross, but that seems to be something that can be cleaned up at the end.  Run the analysis, fill in any gaps with your favorite interpolation.

I dunno, I'm torn; if that's really how various fields perform analysis, I suppose we should support it... although it seems entirely wrong from a theoretical point of view.  But I'm certainly not an expert in this either :)

Regardless, if this functionality is desired, it still seems better to be implemented as part of an `interpolation` or `fit` pipeline as suggested in #24747.  That way individual pipelines don't need to concern themselves with gaps, and we can centralize all the interpolating logic in one place.

**Comment by rjernst** [comment]
> If any kind of interpolating policy is used, a daily derivative will show a change from Friday to Saturday, which is just wrong

I think the point was to skip Sat/Sun since the markets were closed, so a daily derivative would act as if Sat/Sun did not exist, and Fri->Mon is the diff?

**Comment by polyfractal** [comment]
Mmm, I see.  Yeah I misinterpreted, because I was thinking the proposal would "fill in" the empty buckets with the next good value.  So  `[1, 2, null, null, 5, 6]` would become `[1, 2, 5, 5, 5, 6]` which would produce all kinds of oddities. 

But if the proposal is to skip any null/Nan bucket (like it does today), but when it encounters a real value it finds the last usable value (`2` in the above example) instead of just checking the last bucket, that seems reasonable.

I'm not sure how that would work for other pipeline aggs though, such as movavg?  E.g. if your data looks like: `[1, 2, null, null, 5, 6]` and your window is size `3`, what do we do with a policy like that?  Our windows would be:

- `[1,2,null]`
- `[2,null, null]`
- `[null, null, 5]`

In this case it isn't as simple as just finding the last usable value, since we are looking at windows of time.  We could fill in the nulls with the last usable value, but they we're duplicating data.  Or we can interpolate linearly (or whatever) but that's straying into territory of making up data.

Ok, I'm done thinking out loud now :)

**Comment by polyfractal** [comment]
So this was discussed in last week's FixitFriday.  The general consensus is that `insert_zero` isn't a great option, but there also isn't strong enough reason to remove it at the moment.  And we recognize that we don't know enough about potential use-cases to fully remove `gap_policy` at this time either.

I'm going to put a pin in this issue and close it, although we'll remain on the lookout to see if `gap_policy` should be extended with more interpolation schemes (or add an interpolation agg).  It's a tricky issue since we don't want to recreate R inside of ES, but we also want to provide enough general-purpose tools for it to be useful.  So the plan is to do nothing at the moment and continue collecting more user-scenarios to see what we can/should do.

Thanks for everyone's input! :)
